### ISSUE TITLE
Bug: error in effect cleanup function no longer catchable

### ISSUE THREAD

**Issue by eps1lon** [facebook/react#20946]
React version: 17.0.1
TL;DR: Fixed with `skipUnmountedBoundaries = true`

## Steps To Reproduce

1. Throw in an effect cleanup 
2. Try to catch it when wrapped in `act()`

Link to code example:
[React 16](https://codesandbox.io/s/react-16-throw-in-effect-destory-kdeij)
[React 17](https://codesandbox.io/s/react-17-throw-in-effect-destory-forked-5n864)

I also tested it locally and jest and couldn't catch the error as well.

## The current behavior

The error is uncatchable and does not include the component stack ("The above error occurred").

## The expected behavior

The error is catchable and does include the component stack i.e. React 16 behavior.

## Context

I understand that the cleanup is now async but it's still flushed in `act()` so I would expect that an error thrown is catchable.

Originally reported in https://github.com/testing-library/react-testing-library/issues/874

**Comment by eps1lon** [comment]
Added some tests to `ReactTestUtilsAct` and realized this already covered by

**Comment by eps1lon** [comment]
Fixed with https://github.com/facebook/react/pull/23322
Confirmation (notice the caught error during unmount): https://codesandbox.io/s/react-18-throw-in-effect-destory-forked-ce0j2u?file=/src/index.js
